spark-kernel-dsv2/src/main/java/io/delta/spark/dsv2/scan/KernelSparkScanContext.java

```diff
@@ -0,0 +1,112 @@
+import io.delta.spark.dsv2.scan.batch.KernelSparkInputPartition;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.spark.sql.connector.read.InputPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Shared context for Kernel-based Spark scans that manages scan state and partition planning.
+ * Caches FilteredColumnarBatch data and converts them to Spark InputPartitions on demand.
+ */
+public class KernelSparkScanContext {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KernelSparkScanContext.class);
+
+  private final Scan kernelScan;
+  private final Engine engine;
+  private final String serializedScanState;
+  private final AtomicReference<Optional<List<FilteredColumnarBatch>>> cachedBatches;
+
+  public KernelSparkScanContext(Scan kernelScan, Engine engine) {
+    this.kernelScan = requireNonNull(kernelScan, "kernelScan is null");
+    this.engine = requireNonNull(engine, "engine is null");
+    this.serializedScanState = JsonUtils.rowToJson(kernelScan.getScanState(engine));
+    this.cachedBatches = new AtomicReference<>(Optional.empty());
+  }
+
+  /**
+   * Plans and creates input partitions for Spark execution.
+   * 
+   * <p>This method converts Delta Kernel scan files into Spark InputPartitions. Each file row
+   * from the kernel scan creates one input partition containing serialized scan state and file
+   * metadata.
+   * 
+   * <p>Results are cached on first invocation using thread-safe lazy initialization. Subsequent
+   * calls return the partitions calculated based on the same cached batches..
+   * 
+   * @return array of InputPartitions for Spark to process
+   * @throws UncheckedIOException if kernel scan fails or partition creation fails
+   */
+  public InputPartition[] planPartitions() {
+    Optional<List<FilteredColumnarBatch>> batches = cachedBatches.get();
+    if (!batches.isPresent()) {
+      synchronized (this) {
+        batches = cachedBatches.get();
+        if (!batches.isPresent()) {
+          List<FilteredColumnarBatch> batchList = new ArrayList<>();
+          try (CloseableIterator<FilteredColumnarBatch> scanFiles =
+              kernelScan.getScanFiles(engine)) {
+            while (scanFiles.hasNext()) {
+              FilteredColumnarBatch batch = scanFiles.next();
+              batchList.add(batch);
+            }
+          } catch (IOException e) {
+            LOG.warn("Failed to read from kernel scan", e);
+            throw new UncheckedIOException("Failed to read added files from kernel scan", e);
+          }
+          batches = Optional.of(batchList);
+          cachedBatches.set(batches);
+        }
+      }
+    }
+
+    // Create InputPartitions from cached batches
+    // Implementation note: not caching partitions for the future extension of runtime filtering.
```

Comment:
Can we make this note just a bit clearer and expressive? e.g. just elaborate a bit more?